### PR TITLE
Don't reindex an new value on setitem if the original dataframe was empty

### LINKED ISSUE #8023: [BUG] 0.19 cudf empty df assignment issues
**Describe the bug**
Cannot perform previously available functionality against empty dataframe.

**Steps/Code to reproduce bug**
!!!Cudf branch 0.19!!!
```
import cudf
jdf = cudf.DataFrame({"alpha": [1,2,3,3,4,5,6]})
gdf = cudf.DataFrame()
new_cols = [c for c in jdf.columns if c not in gdf.columns]
gdf[new_cols] = jdf[new_cols]
print(gdf)
```
result:
```
Empty DataFrame
Columns: [alpha]
Index: []
```

**Expected behavior**
Expect dataframe gdf to have the column names and values of the jdf dataframe new_cols columns. This change occurred about a week ago.

### PR DESCRIPTION
**Pull request by vyasr**
This PR resolves #8023. The `pandas` compatibility patch in #7948 (and the backport to 0.19 in #7957) also needs to handle the case of assigning to an empty DataFrame.

### CODE REVIEW

**Review by shwina — APPROVED** [review]
LGTM!

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -7945,7 +7945,7 @@ def _setitem_with_dataframe(
             "Number of Input Columns must be same replacement Dataframe"
         )
 
-    if not input_df.index.equals(replace_df.index):
+    if not input_df.index.equals(replace_df.index) and len(input_df) != 0:
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:7948]: Can we add a test to cover this?

Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:7948]: Yes I'm working on it now, will push as soon as it's ready.

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/core/dataframe.py [hunk 2]
```diff
@@ -7953,7 +7953,7 @@ def _setitem_with_dataframe(
...
```
Comment (galipremsagar) [inline python/cudf/cudf/core/dataframe.py:7956]:
```suggestion
    if len(input_df) != 0 and not input_df.index.equals(replace_df.index):
```

**Review by galipremsagar — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/dataframe.py (hunk 2 above)
Comment (kkraus14) [reply python/cudf/cudf/core/dataframe.py:7956]: Good call on short circuiting an expensive `equals` check

**Review by galipremsagar — APPROVED** [review]

**Pull request merged**